addons/website_sale/controllers/main.py

```diff
@@ -1230,3 +1231,69 @@ def products_autocomplete(self, term, options={}, **kwargs):
+    @http.route('/shop/products/recently_viewed', type='json', auth='public', website=True)
+    def products_recently_viewed(self, **kwargs):
+        """
+        Returns list of recently viewed products according to current user and product options
+        """
+        excluded_products = []
+        max_number_of_product_for_carousel = 12
+        current_sale_order = request.website.sale_get_order()
+        if current_sale_order:
+            # Exclude items already in the cart
+            # order_lines = request.env['sale.order.line'].sudo().search([('order_id', '=', current_sale_order.id)])
+            excluded_products += current_sale_order.mapped('line_ids.product_id')
+        if not request.env.user._is_public():
+            partner_id = request.env.user.partner_id
+            # Retrieve all the previous partner's visitors to have full history of his last products viewed.
+            visitors = request.env['website.visitor'].sudo().with_context(active_test=False).search([('user_partner_id', '=', partner_id.id)])
+        else:
+            visitors = request.env['website.visitor']._get_visitor_from_request()
+        if visitors:
+            # Keep only 12 products for the carousel
+            length = max_number_of_product_for_carousel + len(excluded_products)
+            products = visitors.website_track_ids.product_id[:length]
+            if excluded_products:
+                products = (products - excluded_products)[:max_number_of_product_for_carousel]
+            if products:
+                res = {
+                    'products': products.read(['id', 'name', 'website_url']),
+                }
+
+                FieldMonetary = request.env['ir.qweb.field.monetary']
+                monetary_options = {
+                    'display_currency': request.website.get_current_pricelist().currency_id,
+                }
+                rating = request.website.viewref('website_sale.product_comment').active
+                for res_product, product in zip(res['products'], products):
+                    combination_info = product._get_combination_info_variant()
+                    res_product.update(combination_info)
+                    res_product['list_price'] = FieldMonetary.value_to_html(res_product['list_price'], monetary_options)
+                    res_product['price'] = FieldMonetary.value_to_html(res_product['price'], monetary_options)
+                    if rating:
+                        res_product['rating'] = request.env["ir.ui.view"].render_template('website_rating.rating_widget_stars_static', values={
+                            'rating_avg': product.rating_avg,
+                            'rating_count': product.rating_count,
+                        })
+
+                return res
+        return {}
+
+    @http.route('/shop/products/recently_viewed_update', type='json', auth='public', website=True)
+    def products_recently_viewed_update(self, product_id, **kwargs):
```

Comment:
as discuss previously add the debounce to more than 3 sec... 
I let you decide your best delay ;)